File: lab_1.4/trees_1.4.java
// РЕАЛИЗАЦИЯ ДЕРЕВЬЕВ НА JAVA

// 1. Внутренний класс вершины дерева
private class Node {
    public int key;
    public Node left;
    public Node right;

    public Node(int key) {
        this.key = key;
    }
}

// 2. Основной класс дерева
class Tree {
    private static class Node {
        int key;
        Node left;
        Node right;
    }

    private Node root;

    // 3. Метод добавления элемента в дерево
    // Рекурсивная реализация
    public void insert(int x) {
        root = doInsert(root, x);
    }

    private static Node doInsert(Node node, int x) {
        if (node == null) {
            return new Node(x);
        }
        if (x < node.key) {
            node.left = doInsert(node.left, x);
        } else if (x > node.key) {
            node.right = doInsert(node.right, x);
        }
        return node;
    }

    // Нерекурсивная реализация
    public void insertIterative(int x) {
        Node parent = null;
        Node node = root;
        while (node != null) {
            parent = node;
            if (x < node.key) {
                node = node.left;
            } else if (x > node.key) {
                node = node.right;
            } else {
                return; // Элемент уже существует
            }
        }
        Node newNode = new Node(x);
        if (parent == null) {
            root = newNode;
        } else if (x < parent.key) {
            parent.left = newNode;
        } else if (x > parent.key) {
            parent.right = newNode;
        }
    }
}

File: lab_1.4/trees_1.4.py
# АЛГОРИТМЫ РЕАЛИЗАЦИИ ДЕРЕВЬЕВ НА PYTHON

# 1. Класс вершины дерева
class Node(object):
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None


# 2. Класс дерева
class Tree(object):
    def __init__(self):
        self.root = None


# 3. Поиск ключа в дереве

# Рекурсивная реализация
def search_recursively(v, x):
    if v is None or v.key == x:
        return v
    elif x < v.key:
        return search_recursively(v.left, x)
    else:  # x > v.key
        return search_recursively(v.right, x)

# Нерекурсивная реализация
def search_iteratively(root, x):
    v = root
    while v is not None:
        if v.key == x:
            return v
        elif x < v.key:
            v = v.left
        else:  # x > v.key
            v = v.right
    return None


# 4. Добавление ключа в дерево

# Рекурсивная реализация
def insert_recursively(v, x):
    if v is None:
        return Node(x)
    if x < v.key:
        v.left = insert_recursively(v.left, x)
    elif x > v.key:
        v.right = insert_recursively(v.right, x)
    # v.key == x (ничего не делаем, если ключ уже присутствует)
    return v

tree.root = insert_recursively(tree.root, x)

# Нерекурсивная реализация
def insert_iteratively(x):
    parent = None
    v = tree.root
    while v is not None:
        parent = v
        if x < v.key:
            v = v.left
        elif x > v.key:
            v = v.right
        else:  # x == v.key
            return
    w = Node(x)
    if parent is None:
        tree.root = w
    elif x < parent.key:
        parent.left = w
    elif x > parent.key:
        parent.right = w

# 5. Удаление из дерева.

# Рекурсивная реализация
def delete_recursively(v, x):
    if v is None:
        return None
    if x < v.key:
        v.left = delete_recursively(v.left, x)
        return v
    elif x > v.key:
        v.right = delete_recursively(v.right, x)
        return v
    # v.key == x
    if v.left is None:
        return v.right
    elif v.right is None:
        return v.left
    else:
        # Both subtrees are present
        min_key = find_min(v.right).key
        v.key = min_key
        v.right = delete_recursively(v.right, min_key)
        return v

def find_min(v):
    if v.left is not None:
        return find_min(v.left)
    else:
        return v

# Нерекурсивная реализация
def replace_child(parent, old, new):
    if parent is None:
        tree.root = new
    elif parent.left == old:
        parent.left = new
    elif parent.right == old:
        parent.right = new

def delete_iteratively(x):
    parent = None
    v = tree.root
    while True:
        if v is None:
            return
        if x < v.key:
            parent = v
            v = v.left
        elif x > v.key:
            parent = v
            v = v.right
        else:  # x == v.key
            break
    result = None
    if v.left is None:
        result = v.right
    elif v.right is None:
        result = v.left
    else:
        min_node_parent = v
        min_node = v.right
        while min_node.left is not None:
            min_node_parent = min_node
            min_node = min_node.left
        result = v
        v.key = min_node.key
        replace_child(min_node_parent, min_node, min_node.right)
    replace_child(parent, v, result)
